### ISSUE TITLE
dma_declare_coherent_memory() usage in kmd

### ISSUE THREAD

**Issue by robotcn** [nvdla/sw#95]
Hi all, we're bring up nvdla on xilinx fpga with small config, with kernel version 4.14. 
I have a question about the drm dma memory allocation in kmd. Please check following function,

        dma = dma_declare_coherent_memory(drm->dev, 0xC0000000, 0xC0000000,
                        0x40000000, DMA_MEMORY_MAP | DMA_MEMORY_EXCLUSIVE);

In 4.14 kernel, the flag 'DMA_MEMORY_MAP' is removed, so I just remove this flag in code.
I would like to know why the physical and device address parameter is hard code to 0xC0000000. (I found the ram base address is also 0xc0000000 in conf/aarch64_nvdla.lua for vp). In my case, the fpga board ddr address space should be 0 ~ 2GB, then how should I modify this for my board? 

Any reply is appreciated. Thanks.

**Comment by mmaciag** [comment]
Hi, please have a look on these two threads:

https://github.com/nvdla/hw/issues/110
https://github.com/nvdla/sw/issues/70

**Comment by mmaciag** [comment]
Yes, this is fast way - more portable way would be implementing extra device tree node (which I think was suggested in the source comment). Note that in nvdla/hw#110 I 
mistakenly used `nvidia,nvdla_os_initial` in device tree for a while - the correct entry for nv_small is `nvdla,nvdla_2`

**Comment by robotcn** [comment]
@mmaciag  Thanks for your reminder. Just make sure the .compatible property in driver for small config match the .compatible property in device node, the the driver will match the device when probe. And another change for small config is the register definition in opendla.h

By the way, I found the system-top.dts is automatically generated, have you add the reserved-memory node in this file? Or any other source file?

**Comment by ddkkevin** [comment]
Add node in file project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi.

**Comment by robotcn** [comment]
Thanks all. Eventually we get log "Test Pass" on fpga with nvdla small config, regression tests for kmd. 
As ddkkevin said, we add following in 'project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi'. This means to reserve 256MB dma memory from 0x70000000, and used for IP 'NV_nvdla_wrapper_0'.

        reserved-memory {
                #address-cells = <2>;
                #size-cells = <2>;
                ranges;

                nvdla_dma: drm-dma@0 {
                          no-map;
                          reg = <0x0 0x70000000 0x0 0x10000000>;
                };
        };
        @NV_nvdla_wrapper_0 {
                memory-region = <&nvdla_dma>;
        };

**Comment by VVViy** [comment]
hi, @robotcn , we are stuck in installing drm.ko and opendla.ko, can you share your process? I notice your kernel is 4.14 (we use petalinux 2017.4 based on kernel 4.9 ). Did you compile the opendla.ko for your kernel version?  

We appreciate your response. Thanks.

**Comment by VVViy** [comment]
Thank you for your quick reply. We previously intended to build opendla.ko with petalinux src, but when we 'make KDIR' with 'kernel image' in images/linux/, error reported, and we failed to find kernel src dir. Which kernel src you used? 
Thank you very much.

**Comment by robotcn** [comment]
We use 'petalinux-create -t modules --name opendla --enable' to create kernel module, and copy kmd source code to the module, and 'petalinux-build -c opendla' to generate opendla.ko.

**Comment by VVViy** [comment]
Oh, yes, I forgot that, thank you for your reminder! Additionally, one question is still confused me, how did you deal with the 'nvdla2csb_wr_complete' signal in block design? It is not a AXI standerd signal, so 'make external' or another interrupt connected to PS?

**Comment by robotcn** [comment]
Our hw engineer told me he didn't export this signal to PS.

**Comment by VVViy** [comment]
OK，thank you very much！

**Comment by peterzh2018** [comment]
> 
> 
> Thanks all. Eventually we get log "Test Pass" on fpga with nvdla small config, regression tests for kmd.
> As ddkkevin said, we add following in 'project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi'. This means to reserve 256MB dma memory from 0x70000000, and used for IP 'NV_nvdla_wrapper_0'.
> 
> ```
>     reserved-memory {
>             #address-cells = <2>;
>             #size-cells = <2>;
>             ranges;
> 
>             nvdla_dma: drm-dma@0 {
>                       no-map;
>                       reg = <0x0 0x70000000 0x0 0x10000000>;
>             };
>     };
>     @NV_nvdla_wrapper_0 {
>             memory-region = <&nvdla_dma>;
>     };
> ```


> 
> 
> Thanks all. Eventually we get log "Test Pass" on fpga with nvdla small config, regression tests for kmd.
> As ddkkevin said, we add following in 'project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi'. This means to reserve 256MB dma memory from 0x70000000, and used for IP 'NV_nvdla_wrapper_0'.
> 
> ```
>     reserved-memory {
>             #address-cells = <2>;
>             #size-cells = <2>;
>             ranges;
> 
>             nvdla_dma: drm-dma@0 {
>                       no-map;
>                       reg = <0x0 0x70000000 0x0 0x10000000>;
>             };
>     };
>     @NV_nvdla_wrapper_0 {
>             memory-region = <&nvdla_dma>;
>     };
> ```
@robotcn 

I am trying to run the NN_L0_1_small_fbuf test on the NVDLA small on fpga ZCU102
can't run NN_L0_1_small_fbuf? Has anyone had the same problem?
== Tests for nv_small ==

= Run PDP/PDP_L0_0_small_fbuf
creating new runtime context...
Emulator starting
submitting tasks...
Shutdown signal received, exiting
Test pass
= Run CONV/CONV_D_L0_0_small_fbuf
creating new runtime context...
Emulator starting
submitting tasks...
Shutdown signal received, exiting
Test pass
= Run SDP/SDP_X1_L0_0_small_fbuf
creating new runtime context...
Emulator starting
submitting tasks...
Shutdown signal received, exiting
Test pass
= Run CDP/CDP_L0_0_small_fbuf
creating new runtime context...
Emulator starting
submitting tasks...
Shutdown signal received, exiting
Test pass
= Run NN/NN_L0_1_small_fbuf
creating new runtime context...
Emulator starting
submitting tasks

can't continue to run， Did you have the same problem? What may be the problem?

**Comment by xuehui** [comment]
> 
> 
> > Thanks all. Eventually we get log "Test Pass" on fpga with nvdla small config, regression tests for kmd.
> > As ddkkevin said, we add following in 'project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi'. This means to reserve 256MB dma memory from 0x70000000, and used for IP 'NV_nvdla_wrapper_0'.
> > ```
> >     reserved-memory {
> >             #address-cells = <2>;
> >             #size-cells = <2>;
> >             ranges;
> > 
> >             nvdla_dma: drm-dma@0 {
> >                       no-map;
> >                       reg = <0x0 0x70000000 0x0 0x10000000>;
> >             };
> >     };
> >     @NV_nvdla_wrapper_0 {
> >             memory-region = <&nvdla_dma>;
> >     };
> > ```
> 
> > Thanks all. Eventually we get log "Test Pass" on fpga with nvdla small config, regression tests for kmd.
> > As ddkkevin said, we add following in 'project-spec/meta-user/recipes-bsp/device-tree/files/system-user.dtsi'. This means to reserve 256MB dma memory from 0x70000000, and used for IP 'NV_nvdla_wrapper_0'.
> > ```
> >     reserved-memory {
> >             #address-cells = <2>;
> >             #size-cells = <2>;
> >             ranges;
> > 
> >             nvdla_dma: drm-dma@0 {
> >                       no-map;
> >                       reg = <0x0 0x70000000 0x0 0x10000000>;
> >             };
> >     };
> >     @NV_nvdla_wrapper_0 {
> >             memory-region = <&nvdla_dma>;
> >     };
> > ```
> 
> @robotcn
> 
> I am trying to run the NN_L0_1_small_fbuf test on the NVDLA small on fpga ZCU102
> can't run NN_L0_1_small_fbuf? Has anyone had the same problem?
> == Tests for nv_small ==
> 
> = Run PDP/PDP_L0_0_small_fbuf
> creating new runtime context...
> Emulator starting
> submitting tasks...
> Shutdown signal received, exiting
> Test pass
> = Run CONV/CONV_D_L0_0_small_fbuf
> creating new runtime context...
> Emulator starting
> submitting tasks...
> Shutdown signal received, exiting
> Test pass
> = Run SDP/SDP_X1_L0_0_small_fbuf
> creating new runtime context...
> Emulator starting
> submitting tasks...
> Shutdown signal received, exiting
> Test pass
> = Run CDP/CDP_L0_0_small_fbuf
> creating new runtime context...
> Emulator starting
> submitting tasks...
> Shutdown signal received, exiting
> Test pass
> = Run NN/NN_L0_1_small_fbuf
> creating new runtime context...
> Emulator starting
> submitting tasks
> 
> can't continue to run， Did you have the same problem? What may be the problem?

I got the same problem to run NN_L0_1_small_fbuf.  It stuck in sdp , no work done interrupt.
